retrieval/k8s/k8s_stats.go

```diff
@@ -1,82 +1,102 @@
-			return err
-		}
+	// v1Sources
+	replicationControllerInformer := factory.Core().V1().ReplicationControllers().Informer()
+	servicesInformer := factory.Core().V1().Services().Informer()
+	nodesInformer := factory.Core().V1().Nodes().Informer()
+	podsInformer := factory.Core().V1().Pods().Informer()
+	persistentVolumesInformer := factory.Core().V1().PersistentVolumes().Informer()
+	persistentVolumeClaimsInformer := factory.Core().V1().PersistentVolumeClaims().Informer()
+	namespacesInformer := factory.Core().V1().Namespaces().Informer()
+	// AppSources
+	replicasetsInformer := factory.Apps().V1().ReplicaSets().Informer()
+	daemonsetsInformer := factory.Apps().V1().DaemonSets().Informer()
+	deploymentsInformer := factory.Apps().V1().Deployments().Informer()
+	// Jobs
+	jobsInformer := factory.Batch().V1().Jobs().Informer()
+	// Cronjobs were introduced in k8s 1.21 so for older versions do not attempt to create an informer
+	var cronJobsInformer cache.SharedIndexInformer
+	if clusterVersion > 1.20 {
+		cronJobsInformer = factory.Batch().V1().CronJobs().Informer()
 	}
+	// closing this will kill all informers
+	stopCh := make(chan struct{})
+	// runs in background, starts all informers that are a part of the factory
+	factory.Start(stopCh)
+	// wait until all informers have successfully synced
+	factory.WaitForCacheSync(stopCh)
 
-	// get v1beta1 sources
-	for _, v1b1s := range v1beta1Sources {
-		_, err := rawClient.GetRawEndPoint(
-			http.MethodGet, v1b1s, workDir, clusterHostURL+"/apis/extensions/v1beta1/"+v1b1s, nil, true)
-		if err != nil {
-			log.Errorf("Error retrieving "+v1b1s+" metric endpoint: %s", err)
-			return err
-		}
+	var clusterInformers = map[string]*cache.SharedIndexInformer{
+		"replicationcontrollers": &replicationControllerInformer,
+		"services":               &servicesInformer,
+		"nodes":                  &nodesInformer,
+		"pods":                   &podsInformer,
+		"persistentvolumes":      &persistentVolumesInformer,
+		"persistentvolumeclaims": &persistentVolumeClaimsInformer,
+		"replicasets":            &replicasetsInformer,
+		"daemonsets":             &daemonsetsInformer,
+		"deployments":            &deploymentsInformer,
+		"namespaces":             &namespacesInformer,
+		"jobs":                   &jobsInformer,
+		"cronjobs":               &cronJobsInformer,
 	}
+	return clusterInformers, nil
+}
 
-	// get v1 App sources
-	for _, v1Apps := range v1AppSources {
-		_, err := rawClient.GetRawEndPoint(http.MethodGet, v1Apps, workDir, clusterHostURL+"/apis/apps/v1/"+v1Apps, nil, true)
+//GetK8sMetricsFromInformer loops through all k8s resource informers in kubeAgentConfig writing each to the WSD
+func GetK8sMetricsFromInformer(informers map[string]*cache.SharedIndexInformer, workDir *os.File) error {
+	for resourceName, informer := range informers {
+		// Cronjob informer will be nil if k8s version is less than 1.21, if so skip getting the list of cronjobs
+		if *informer == nil {
```

Comment:
If you do a `*object` when object is nil, go will panic. If there's an easy way to inject logs here or when the informers are created to have some information on what exactly is being gathered that could be good as well for troubleshooting purposes.

```suggestion
		if informer == nil {
```

Reply:
That is super interesting, because when I tried your suggestion, informer was a memory location. Then when I tested it with the dereference I got the nil value.